powersimdata/network/europe_tub/model.py

```diff
@@ -0,0 +1,140 @@
+import os
+import shutil
+from zipfile import ZipFile
+
+from powersimdata.network.constants.region.europe import (
+    abv2country,
+    abv2timezone,
+    interconnect2abv,
+)
+from powersimdata.network.helpers import (
+    check_and_format_interconnect,
+    interconnect_to_name,
+)
+from powersimdata.network.model import ModelImmutables
+from powersimdata.utility.helpers import _check_import
+
+pypsa = _check_import("pypsa")
+zenodo_get = _check_import("zenodo_get")
+
+
+class TUB:
+    """PyPSA Europe network.
+
+    :param str/iterable interconnect: interconnect name(s).
+    :param int reduction: reduction parameter (number of nodes in network). If None,
+        the full network is loaded.
+    :param bool overwrite: the existing dataset is deleted a new dataset is downloaded
+        from zenodo.
+    """
+
+    def __init__(self, interconnect, reduction=None, overwrite=False):
+        """Constructor."""
+        self.grid_model = "europe_tub"
+        self.interconnect = check_and_format_interconnect(
+            interconnect, model=self.grid_model
+        )
+        self.data_loc = os.path.join(os.path.dirname(__file__), "data")
+        self.zenodo_record_id = "3601881"
+        self.reduction = reduction
+
+        if overwrite:
+            self.remove_data()
+
+        self.retrieve_data()
+
+    def remove_data(self):
+        """Remove data stored on disk"""
+        print("Removing PyPSA-Eur dataset")
+        shutil.rmtree(self.data_loc)
+
+    def retrieve_data(self):
+        """Fetch data"""
+        zenodo_get.zenodo_get([self.zenodo_record_id, "-o", f"{self.data_loc}"])
+        with ZipFile(os.path.join(self.data_loc, "networks.zip"), "r") as zip_network:
+            zip_network.extractall(self.data_loc)
+
+    def build(self):
+        """Build network"""
+        path = os.path.join(self.data_loc, "networks")
+        if self.reduction is None:
+            network = pypsa.Network(os.path.join(path, "elec_s.nc"))
+        elif os.path.exists(os.path.join(path, f"elec_s_{self.reduction}.nc")):
+            network = pypsa.Network(os.path.join(path, f"elec_s_{self.reduction}.nc"))
+        else:
+            raise ValueError(
+                "Invalid Resolution. Choose among: None | 1024 | 512 | 256 | 128 | 37"
+            )
+        id2zone = {i: l for i, l in enumerate(network.buses.index)}
+        zone2id = {l: i for i, l in id2zone.items()}
+
+        if "Europe" in self.interconnect:
+            self.network = network
```

Comment:
So we will have `self.network` which contains all the tables instead of `self.plant`, `self.branch`, `self.bus`, etc in other grid models. How this would impact the downstream functions in `PostREISE` and will this be compatible with [export_to_pypsa() ](https://github.com/Breakthrough-Energy/PowerSimData/blob/5682629dec9f8b7470bf32335a7f74b2310df076/powersimdata/input/export_data.py#L289)?

Reply:
It is up to discussion. Where do we want to convert the `Network` object to a `Grid` object? In PR #632 where I refactor the `_Builder` to have a base class (`_Builder`) and two child classes (`FromCSV` and from `PyPSA`), see: https://github.com/Breakthrough-Energy/PowerSimData/blob/67902e65440daf20cc18adf92af1b3e24544bb83/powersimdata/scenario/create.py#L419 we get the `Grid` and the profiles from the `Network`, and then we can populate the change table and transform the grid and profiles accordingly. The workflow is:
```
from powersimdata import Scenario

scenario = Scenario()

scenario.set_grid(grid_model="europe_tub", interconnect="Nordic")

scenario.set_name("test", "dummy")
scenario.set_time("2016-08-01 00:00:00", "2016-08-31 23:00:00", "24H")
scenario.change_table.scale_plant_capacity(...)
```

Reply:
I see. I didn't look into #635 yet. So the higher level workflow when a user wants to run a `europe_tub` model would be: a) instantiate a scenario and set the grid model to be `europe_tub` b) add modifications to change table c) run with PyPSA engine (CEM) (in this case, we will need to convert the scenario object back to a network object upon feeding into PyPSA) or REISE.jl engine (PCM). Then in step a) we will instantiate a grid object from a network object. The grid is built via `FromCSV` when the model is `TAMU` or `HIFLD` and via `FromPyPSA` when the model is `TUB`. Given the implementation of `TUB` (the build function) here, when instantiating the grid object based on it, we will need `tub_network = TUB(interconnect) self.bus = tub_network.network.buses` instead of `hifld_network = HIFLD(interconnect) self.bus = hifld_network.bus`, right?

Reply:
We will do as usual for usa_tamu (and hifld) and for europe_tub, we will create a ˋTUB` object, build the network from zenodo's NetCDF files (for now) via ˋtub = TUB(interconnect)` and ˋtub.build()` and pass ˋtub.network` to the constructor of the ˋFromPyPSAˋ class in #608 (I realize this class has the same name than the child class of the builder) to get a grid object. Not sure how we will get the profile yet.

Reply:
> We will do as usual for usa_tamu (and hifld) and for europe_tub, we will create a ˋTUB` object, build the network from zenodo's NetCDF files (for now) via ˋtub = TUB(interconnect)` and ˋtub.build()` and pass ˋtub.network` to the constructor of the ˋFromPyPSAˋ class in #608 (I realize this class has the same name than the child class of the builder) to get a grid object. Not sure how we will get the profile yet.

That makes perfect sense. Thanks!